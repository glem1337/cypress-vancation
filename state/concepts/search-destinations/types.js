const namespace = 'search-destinations';

export const SEARCH_DESTINATION = `${namespace}/SEARCH_DESTINATION`;
export const SEARCH_DESTINATIONS_BY_QUERY = `${namespace}/SEARCH_DESTINATIONS_BY_QUERY`;
export const SEARCH_DESTINATIONS_BY_COORDINATES = `${namespace}/SEARCH_DESTINATIONS_BY_COORDINATES`;
export const SET_DESTINATIONS_BY_COORDS_IDS = `${namespace}/SET_DESTINATIONS_BY_COORDS_IDS`;
export const SET_DESTINATIONS_BY_QUERY_IDS = `${namespace}/SET_DESTINATIONS_BY_QUERY_IDS`;
export const SET_DESTINATIONS_FROM_MAP_BOX_IDS = `${namespace}/SET_DESTINATIONS_FROM_MAP_BOX_IDS`;
export const CLEAR_DESTINATIONS_BY_QUERY_IDS = `${namespace}/CLEAR_DESTINATIONS_BY_QUERY_IDS`;
export const SET_SEARCH_DESTINATION_DATES = `${namespace}/SET_SEARCH_DESTINATION_DATES`;
export const SET_SEARCH_DESTINATION_LOCATION = `${namespace}/SET_SEARCH_DESTINATION_LOCATION`;
export const SET_SEARCH_DESTINATION_LOCATION_INTENT = `${namespace}/SET_SEARCH_DESTINATION_LOCATION_INTENT`;
export const SET_SEARCH_DESTINATION_ALL_LOCATIONS = `${namespace}/SET_SEARCH_DESTINATION_ALL_LOCATIONS`;
export const CLEAR_SEARCH_DESTINATION_PARAMS = `${namespace}/CLEAR_SEARCH_DESTINATION_PARAMS`;
export const SET_ACTIVE_CAMPER_ID = `${namespace}/SET_ACTIVE_CAMPER_ID`;
export const FETCH_SEARCH_RESULT_DATA = `${namespace}/FETCH_SEARCH_RESULT_DATA`;
// Filters
export const SET_FILTER_VALUE = `${namespace}/SET_FILTER_VALUE`;
export const SET_FILTER_BATCH_VALUE = `${namespace}/SET_FILTER_BATCH_VALUE`;
export const CLEAR_ALL_FILTERS = `${namespace}/CLEAR_ALL_FILTERS`;
export const TOGGLE_MOBILE_FILTERS_VISIBILITY = `${namespace}/TOGGLE_MOBILE_FILTERS_VISIBILITY`;
export const TOGGLE_DESKTOP_MAP_VISIBILITY = `${namespace}/TOGGLE_DESKTOP_MAP_VISIBILITY`;
// Near you destinations
export const FETCH_NEAR_YOU_DESTINATIONS = `${namespace}/FETCH_NEAR_YOU_DESTINATIONS`;
export const SET_NEAR_YOU_DESTINATIONS_IDS = `${namespace}/SET_NEAR_YOU_DESTINATIONS_IDS`;
