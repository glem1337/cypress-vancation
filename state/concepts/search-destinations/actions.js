import {
  SEARCH_DESTINATION,
  SEARCH_DESTINATIONS_BY_COORDINATES,
  SEARCH_DESTINATIONS_BY_QUERY,
  SET_DESTINATIONS_BY_COORDS_IDS,
  SET_DESTINATIONS_BY_QUERY_IDS,
  SET_DESTINATIONS_FROM_MAP_BOX_IDS,
  CLEAR_DESTINATIONS_BY_QUERY_IDS,
  SET_SEARCH_DESTINATION_DATES,
  SET_SEARCH_DESTINATION_LOCATION,
  CLEAR_SEARCH_DESTINATION_PARAMS,
  SET_ACTIVE_CAMPER_ID,
  SET_FILTER_VALUE,
  SET_FILTER_BATCH_VALUE,
  CLEAR_ALL_FILTERS,
  SET_SEARCH_DESTINATION_LOCATION_INTENT,
  TOGGLE_MOBILE_FILTERS_VISIBILITY,
  TOGGLE_DESKTOP_MAP_VISIBILITY,
  FETCH_NEAR_YOU_DESTINATIONS,
  SET_NEAR_YOU_DESTINATIONS_IDS,
  FETCH_SEARCH_RESULT_DATA,
  SET_SEARCH_DESTINATION_ALL_LOCATIONS,
} from './types';

export const searchDestination = () => ({
  type: SEARCH_DESTINATION,
});

export const searchDestinationsByQuery = (query) => ({
  type: SEARCH_DESTINATIONS_BY_QUERY,
  query,
});

export const searchDestinationsByCoordinates = ({ latitude, longitude }) => ({
  type: SEARCH_DESTINATIONS_BY_COORDINATES,
  latitude,
  longitude,
});

export const setDestinationsByCoordinatesIds = (ids) => ({
  type: SET_DESTINATIONS_BY_COORDS_IDS,
  ids,
});

export const setDestinationsByQueryIds = (ids) => ({
  type: SET_DESTINATIONS_BY_QUERY_IDS,
  ids,
});

export const setDestinationsMapBoxIds = (ids) => ({
  type: SET_DESTINATIONS_FROM_MAP_BOX_IDS,
  ids,
});

export const clearDestinationsByQueryIds = () => ({
  type: CLEAR_DESTINATIONS_BY_QUERY_IDS,
});

export const setSearchDestinationDates = (dateRange) => ({
  type: SET_SEARCH_DESTINATION_DATES,
  dateRange,
});

export const setSearchDestinationLocation = (location) => ({
  type: SET_SEARCH_DESTINATION_LOCATION,
  location,
});

export const clearSearchDestinationParams = () => ({
  type: CLEAR_SEARCH_DESTINATION_PARAMS,
});

export const setActiveCamperId = (id) => ({
  type: SET_ACTIVE_CAMPER_ID,
  id,
});

export const setFilterValue = ({ name, value }) => ({
  type: SET_FILTER_VALUE,
  name,
  value,
});

export const setFilterBatchValue = (filters) => ({
  type: SET_FILTER_BATCH_VALUE,
  filters,
});

export const clearAllFilter = () => ({
  type: CLEAR_ALL_FILTERS,
});

export const setSearchDestinationLocationIntent = (location) => ({
  type: SET_SEARCH_DESTINATION_LOCATION_INTENT,
  location,
});

export const setSearchDestinationAllLocations = (location) => ({
  type: SET_SEARCH_DESTINATION_ALL_LOCATIONS,
  location,
});

export const toggleMobileFiltersVisibility = () => ({
  type: TOGGLE_MOBILE_FILTERS_VISIBILITY,
});

export const toggleDesktopMapVisibility = () => ({
  type: TOGGLE_DESKTOP_MAP_VISIBILITY,
});

export const fetchNearYouDestinations = () => ({
  type: FETCH_NEAR_YOU_DESTINATIONS,
});

export const setNearYouDestinationsIds = (ids) => ({
  type: SET_NEAR_YOU_DESTINATIONS_IDS,
  ids,
});

export const fetchSearchResultData = (placeId) => ({
  type: FETCH_SEARCH_RESULT_DATA,
  placeId,
});
