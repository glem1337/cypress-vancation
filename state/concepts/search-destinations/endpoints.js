import endpoint from 'utils/endpoint';
import {
  searchLandingPagesRoute,
  searchNearestLandingPagesRoute,
  geocodingRoute,
} from 'lib/apiRoutes';

import {
  SEARCH_DESTINATIONS_BY_QUERY,
  SEARCH_DESTINATIONS_BY_COORDINATES,
  FETCH_SEARCH_RESULT_DATA,
} from './types';

export const searchDestinationsByQueryEndpoint = endpoint(SEARCH_DESTINATIONS_BY_QUERY, 'GET', searchLandingPagesRoute);
export const searchDestinationsByCoordinatesEndpoint = endpoint(SEARCH_DESTINATIONS_BY_COORDINATES, 'GET', searchNearestLandingPagesRoute);
export const fetchSearchResultDataOperationEndpoint = placeId => endpoint(FETCH_SEARCH_RESULT_DATA, 'GET', geocodingRoute(placeId));
