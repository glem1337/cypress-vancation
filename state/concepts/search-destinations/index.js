import searchDestinationsReducer from 'state/concepts/search-destinations/reducer';

export { default as searchDestinationsOperations } from './operations';
export * as searchDestinationsTypes from './types';

export default searchDestinationsReducer;
