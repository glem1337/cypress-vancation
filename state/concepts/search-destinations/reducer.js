import { combineReducers } from 'redux';
import { REHYDRATE } from 'redux-persist';
import { v4 as uuid } from 'uuid';
import * as R from 'ramda';
import { HYDRATE } from 'next-redux-wrapper';

import isPresent from 'utils/isPresent';

import {
  SET_DESTINATIONS_BY_COORDS_IDS,
  SET_DESTINATIONS_BY_QUERY_IDS,
  SET_DESTINATIONS_FROM_MAP_BOX_IDS,
  CLEAR_DESTINATIONS_BY_QUERY_IDS,
  SET_SEARCH_DESTINATION_DATES,
  SET_SEARCH_DESTINATION_LOCATION,
  CLEAR_SEARCH_DESTINATION_PARAMS,
  SET_ACTIVE_CAMPER_ID,
  SET_FILTER_VALUE,
  CLEAR_ALL_FILTERS,
  SET_FILTER_BATCH_VALUE,
  SET_SEARCH_DESTINATION_LOCATION_INTENT,
  TOGGLE_MOBILE_FILTERS_VISIBILITY,
  TOGGLE_DESKTOP_MAP_VISIBILITY,
  SET_NEAR_YOU_DESTINATIONS_IDS,
  SET_SEARCH_DESTINATION_ALL_LOCATIONS,
} from './types';

const destinationsByCoordsIds = (state = [], action) => {
  switch (action.type) {
    case SET_DESTINATIONS_BY_COORDS_IDS:
      return action.ids;
    default:
      return state;
  }
};

const destinationsByQueryIds = (state = [], action) => {
  switch (action.type) {
    case SET_DESTINATIONS_BY_QUERY_IDS:
      return action.ids;
    case CLEAR_DESTINATIONS_BY_QUERY_IDS:
      return [];
    default:
      return state;
  }
};

const destinationsFromMapBoxIds = (state = [], action) => {
  switch (action.type) {
    case SET_DESTINATIONS_FROM_MAP_BOX_IDS:
      return action.ids;
    case CLEAR_DESTINATIONS_BY_QUERY_IDS:
      return [];
    default:
      return state;
  }
};

const desktopMapVisibility = (state = true, action) => {
  switch (action.type) {
    case TOGGLE_DESKTOP_MAP_VISIBILITY:
      return !state;
    default:
      return state;
  }
};

const destinationsNearYouIds = (state = [], action) => {
  switch (action.type) {
    case SET_NEAR_YOU_DESTINATIONS_IDS:
      return action.ids;
    default:
      return state;
  }
};

export const searchDestinationParamsInitialState = {
  uuid: null,
  dateRange: null,
  location: null,
  locationIntent: null,
};
const searchDestinationParams = (state = searchDestinationParamsInitialState, action) => {
  switch (action.type) {
    case SET_SEARCH_DESTINATION_DATES:
      return {
        ...state,
        dateRange: action.dateRange,
        uuid: uuid(),
      };
    case SET_SEARCH_DESTINATION_LOCATION:
      return {
        ...state,
        location: action.location,
        uuid: uuid(),
      };
    case SET_SEARCH_DESTINATION_LOCATION_INTENT:
      return {
        ...state,
        locationIntent: action.location,
      };
    case SET_SEARCH_DESTINATION_ALL_LOCATIONS: {
      return {
        ...state,
        location: action.location,
        locationIntent: action.location,
        uuid: uuid(),
      };
    }
    case CLEAR_SEARCH_DESTINATION_PARAMS:
      return searchDestinationParamsInitialState;
    case HYDRATE: {
      const location = R.pathOr(null, ['payload', 'searchDestinations', 'searchDestinationParams', 'location'], action);
      const locationIntent = R.pathOr(null, ['payload', 'searchDestinations', 'searchDestinationParams', 'locationIntent'], action);

      return {
        ...state,
        location,
        locationIntent,
      };
    }
    case REHYDRATE: {
      if (action.payload?.searchDestinationParams) {
        const dates = action.payload?.searchDestinationParams?.dateRange;
        const dateRange = isPresent(dates)
          ? [new Date(dates[0]), new Date(dates[1])]
          : state.dateRange;

        const location = R.isNil(state.location)
          ? action.payload?.searchDestinationParams?.location
          : state.location;

        const locationIntent = R.isNil(state.locationIntent)
          ? action.payload?.searchDestinationParams?.locationIntent
          : state.locationIntent;

        return {
          ...state,
          location,
          locationIntent,
          dateRange,
        };
      }

      return state;
    }
    default:
      return state;
  }
};

export const searchDestinationFiltersInitialState = {
  isMobileVisible: false,
  uuid: null,
  seats: null,
  sleeps: null,
  priceStart: null,
  priceEnd: null,
  vehicles: null,
  delivery: null,
  glamper: null,
  rating: null,
  allowPets: null,
  allowSmoking: null,
  festivalApproved: null,
  allowUnlimitedMiles: null,
  insideHeight: null,
  standardAmenities: null,
  luxuryAmenities: null,
};
const searchDestinationFilters = (state = searchDestinationFiltersInitialState, action) => {
  switch (action.type) {
    case TOGGLE_MOBILE_FILTERS_VISIBILITY: {
      const clone = R.clone(state);

      clone.isMobileVisible = !state.isMobileVisible;

      return clone;
    }
    case SET_FILTER_VALUE: {
      const clone = R.clone(state);

      clone[action.name] = action.value;
      clone.uuid = uuid();

      return clone;
    }
    case SET_FILTER_BATCH_VALUE: {
      const clone = R.clone(state);

      action.filters.forEach(filter => {
        clone[filter.name] = filter.value;
      });
      clone.uuid = uuid();

      return clone;
    }
    case CLEAR_ALL_FILTERS: {
      return {
        ...searchDestinationFiltersInitialState,
        isMobileVisible: state.isMobileVisible,
        uuid: uuid(),
      };
    }
    default:
      return state;
  }
};

const activeCamperId = (state = null, action) => {
  switch (action.type) {
    case SET_ACTIVE_CAMPER_ID:
      return action.id;
    default:
      return state;
  }
};

export default combineReducers({
  destinationsByCoordsIds,
  destinationsByQueryIds,
  destinationsFromMapBoxIds,
  searchDestinationParams,
  searchDestinationFilters,
  activeCamperId,
  desktopMapVisibility,
  destinationsNearYouIds,
});
