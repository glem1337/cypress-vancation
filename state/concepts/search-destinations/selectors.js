import { createSelector } from 'reselect';
import build from 'redux-object';
import * as R from 'ramda';

import isPresent from 'utils/isPresent';

const dataSelector = R.prop('data');
export const destinationsByCoordsIdsSelector = R.path(['searchDestinations', 'destinationsByCoordsIds']);
export const destinationsByQueryIdsSelector = R.path(['searchDestinations', 'destinationsByQueryIds']);
export const destinationsFromMapBoxIdsSelector = R.path(['searchDestinations', 'destinationsFromMapBoxIds']);
export const searchDestinationParamsSelector = R.path(['searchDestinations', 'searchDestinationParams']);
export const searchDestinationFiltersSelector = R.path(['searchDestinations', 'searchDestinationFilters']);
export const activeCamperIdSelector = R.path(['searchDestinations', 'activeCamperId']);
export const desktopMapVisibilitySelector = R.path(['searchDestinations', 'desktopMapVisibility']);
export const destinationsNearYouIdsSelector = R.path(['searchDestinations', 'destinationsNearYouIds']);

export const destinationsByCoordsSelector = createSelector(
  dataSelector,
  destinationsByCoordsIdsSelector,
  (data, ids) => {
    if (R.isEmpty(ids)) {
      return [];
    }

    return R.compose(
      R.filter(item => isPresent(item)),
      R.defaultTo([]),
    )(build(data, 'searchLandingPage', ids));
  },
);

export const destinationsByQuerySelector = createSelector(
  dataSelector,
  destinationsByQueryIdsSelector,
  (data, ids) => {
    if (R.isEmpty(ids)) {
      return [];
    }

    return R.compose(
      R.filter(item => isPresent(item)),
      R.defaultTo([]),
    )(build(data, 'searchLandingPage', ids));
  },
);

export const destinationsFromMapBoxSelector = createSelector(
  dataSelector,
  destinationsFromMapBoxIdsSelector,
  (data, ids) => {
    if (R.isEmpty(ids)) {
      return [];
    }

    return R.compose(
      R.filter(item => isPresent(item)),
      R.defaultTo([]),
    )(build(data, 'mapboxFeature', ids));
  },
);

export const destinationsNearYouSelector = createSelector(
  dataSelector,
  destinationsNearYouIdsSelector,
  (data, ids) => {
    if (R.isEmpty(ids)) {
      return [];
    }

    return R.compose(
      R.filter(item => isPresent(item)),
      R.defaultTo([]),
    )(build(data, 'searchLandingPage', ids));
  },
);
