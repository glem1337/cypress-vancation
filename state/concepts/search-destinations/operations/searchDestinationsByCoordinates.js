import { createLogic } from 'redux-logic';
import normalize from 'json-api-normalizer';
import { pluck } from 'ramda';
import axios from 'axios';
import * as R from 'ramda';
import { v4 as uuid } from 'uuid';

import { ACCESS_TOKEN_MAPBOX } from 'constants/mapbox';
import { SEARCH_DESTINATIONS_BY_COORDINATES_COUNT, MAPBOX_FEATURE_TYPE, SEARCH_RESULTS_RADIUS } from 'constants/searchDestinations';
import { setCurrentCoordinates, setCurrentLocation } from 'state/app/actions';
import { dataApiRequest, dataApiSuccess, dataApiFailure } from 'state/data/actions';
import { reverseGeocodingRoute } from 'lib/apiRoutes';
import getStateShortCodeFromMapboxResult from 'utils/destinations/getStateShortCodeFromMapboxResult';

import { searchDestinationsByCoordinatesEndpoint } from '../endpoints';
import { SEARCH_DESTINATIONS_BY_COORDINATES } from '../types';
import { setDestinationsByCoordinatesIds } from '../actions';

const searchDestinationsByCoordinates = createLogic({
  type: SEARCH_DESTINATIONS_BY_COORDINATES,
  latest: true,

  async process({ httpClient, action }, dispatch, done) {
    const { endpoint, url } = searchDestinationsByCoordinatesEndpoint;

    const params = {
      latitude: action.latitude,
      longitude: action.longitude,
      records_amount: SEARCH_DESTINATIONS_BY_COORDINATES_COUNT,
    };

    dispatch(dataApiRequest({ endpoint }));

    try {
      const { data } = await httpClient.get(url, { params });

      const response = normalize(data);
      dispatch(dataApiSuccess({ response, endpoint }));

      const latitude = parseFloat(data.meta.latitude);
      const longitude = parseFloat(data.meta.longitude);

      const mapboxUrl = reverseGeocodingRoute({
        latitude,
        longitude,
      });

      // Data from MapBox
      const mapBoxResponse = await axios.get(mapboxUrl, {
        params: {
          access_token: ACCESS_TOKEN_MAPBOX,
        },
      });

      // Current user location
      const mapBoxDestination = R.compose(
        item => {
          const shortStateCode = getStateShortCodeFromMapboxResult(item);

          return {
              ...item,
              id: uuid(),
              placeId: item.id,
              latitude: item.center[1],
              longitude: item.center[0],
              shortStateCode,
              placeName: item.place_name,
              placeShortName: item.text,
              type: MAPBOX_FEATURE_TYPE,
              searchRadius: SEARCH_RESULTS_RADIUS,
          };
        },
        R.defaultTo({}),
        R.head,
        R.filter(item => R.test(/(place\.|region\.)/, item.id)),
        R.defaultTo([]),
        R.path(['data', 'features']),
      )(mapBoxResponse);

      dispatch(setCurrentCoordinates({
        latitude,
        longitude,
      }));

      dispatch(setCurrentLocation(mapBoxDestination));

      dispatch(setDestinationsByCoordinatesIds(pluck('id', data.data)));
    } catch (error) {
      dispatch(dataApiFailure({ endpoint, error }));
    }
    done();
  },
});

export default searchDestinationsByCoordinates;
