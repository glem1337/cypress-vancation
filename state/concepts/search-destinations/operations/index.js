import searchDestinationsByQuery from './searchDestinationsByQuery';
import searchDestinationsByCoordinates from './searchDestinationsByCoordinates';
import searchDestination from './searchDestination';
import fetchNearYouDestinations from './fetchNearYouDestinations';
import fetchSearchResultData from './fetchSearchResultData';

export default [
  searchDestinationsByQuery,
  searchDestinationsByCoordinates,
  searchDestination,
  fetchNearYouDestinations,
  fetchSearchResultData,
];
