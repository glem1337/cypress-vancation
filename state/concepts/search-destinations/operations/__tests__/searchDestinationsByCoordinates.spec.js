import normalize from 'json-api-normalizer';
import { pluck } from 'ramda';
import * as R from 'ramda';
import { v4 as uuid } from 'uuid';

import { SEARCH_DESTINATIONS_BY_COORDINATES_COUNT, MAPBOX_FEATURE_TYPE, SEARCH_RESULTS_RADIUS } from 'constants/searchDestinations';
import { searchDestinationsByCoordinatesEndpoint } from 'state/concepts/search-destinations/endpoints';
import { dataApiRequest, dataApiSuccess, dataApiFailure } from 'state/data/actions';
import mockHttpClient from 'utils/testHelpers/mockHttpClient';
import response from 'state/concepts/search-destinations/__mocks__/searchDestinationsByCoordinatesResponse';
import { setCurrentCoordinates, setCurrentLocation } from 'state/app/actions';
import mockedMapboxResponse from 'state/concepts/search-destinations/__mocks__/mapboxSearchResponse';
import getStateShortCodeFromMapboxResult from 'utils/destinations/getStateShortCodeFromMapboxResult';

import { setDestinationsByCoordinatesIds } from '../../actions';

import searchDestinationsByCoordinates from '../searchDestinationsByCoordinates';

jest.mock('axios', () => ({
  get: jest.fn(() => mockedMapboxResponse),
}));

describe('searchDestinationsByCoordinates', () => {
  let dispatch;

  const action = {
    latitude: 'latitude',
    longitude: 'longitude',
  };

  const params = {
    latitude: action.latitude,
    longitude: action.longitude,
    records_amount: SEARCH_DESTINATIONS_BY_COORDINATES_COUNT,
  };

  const { endpoint, url } = searchDestinationsByCoordinatesEndpoint;

  const beforeFunction = httpClient => () => {
    dispatch = jest.fn();
    searchDestinationsByCoordinates.process(
      { httpClient, action, getState: jest.fn() }, dispatch, jest.fn(),
    );
  };

  it('has valid attributes', () => {
    expect(searchDestinationsByCoordinates).toMatchSnapshot();
  });

  describe('success', () => {
    const httpClient = mockHttpClient({ method: 'get', response });

    beforeEach(beforeFunction(httpClient));

    it('calls right endpoint', () => {
      expect(httpClient.get).toHaveBeenCalledWith(url, { params });
    });

    it('success', () => {
      expect(dispatch).toHaveBeenCalledTimes(5);

      expect(dispatch).toHaveBeenNthCalledWith(
        1,
        dataApiRequest({ endpoint }),
      );

      expect(dispatch).toHaveBeenNthCalledWith(
        2,
        dataApiSuccess({
          endpoint,
          response: normalize(response.data),
        }),
      );

      expect(dispatch).toHaveBeenNthCalledWith(
        3,
        setCurrentCoordinates({
          latitude: parseFloat(response.data.meta.latitude),
          longitude: parseFloat(response.data.meta.longitude),
        }),
      );

      // Current user location
      const mapBoxDestination = R.compose(
        item => {
          const shortStateCode = getStateShortCodeFromMapboxResult(item);

          return {
              ...item,
              id: uuid(),
              placeId: item.id,
              latitude: item.center[1],
              longitude: item.center[0],
              shortStateCode,
              placeName: item.place_name,
              placeShortName: item.text,
              type: MAPBOX_FEATURE_TYPE,
              searchRadius: SEARCH_RESULTS_RADIUS,
          };
        },
        R.defaultTo({}),
        R.head,
        R.filter(item => R.test(/(place\.|region\.)/, item.id)),
        R.defaultTo([]),
        R.path(['data', 'features']),
      )(mockedMapboxResponse);

      expect(dispatch).toHaveBeenNthCalledWith(
        4,
        setCurrentLocation(mapBoxDestination),
      );

      expect(dispatch).toHaveBeenNthCalledWith(
        5,
        setDestinationsByCoordinatesIds(pluck('id', response.data.data)),
      );
    });
  });

  describe('failure', () => {
    const error = new Error('test');

    const httpClient = mockHttpClient({ method: 'get', reject: true, response: error });

    beforeEach(beforeFunction(httpClient));

    it('dispatches actions', () => {
      expect(dispatch).toHaveBeenCalledTimes(2);

      expect(dispatch).toHaveBeenNthCalledWith(
        1,
        dataApiRequest({ endpoint }),
      );

      expect(dispatch).toHaveBeenNthCalledWith(
        2,
        dataApiFailure({ endpoint, error }),
      );
    });
  });
});
