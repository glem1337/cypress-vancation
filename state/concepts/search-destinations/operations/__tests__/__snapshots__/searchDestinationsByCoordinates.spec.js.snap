// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`searchDestinationsByCoordinates has valid attributes 1`] = `
Object {
  "cancelType": undefined,
  "debounce": 0,
  "latest": true,
  "name": undefined,
  "process": [Function],
  "processOptions": Object {
    "dispatchMultiple": true,
  },
  "throttle": 0,
  "transform": undefined,
  "type": "search-destinations/SEARCH_DESTINATIONS_BY_COORDINATES",
  "validate": undefined,
  "warnTimeout": 60000,
}
`;
